### PR TITLE
Hierarchical Pages: Functionality to sort a list of page objects and add indent levels

### PR DESCRIPTION
**Pull request by jblz**
This PR adds a helper function that sorts pages hierarchically.

The tests added as part of the commit verify desired behaviors:
* children are placed under parents
* sorting is first done by hierarchy, then by `menu_order`
* orphaned children are placed at top-level and their children are beneath them

To test:

```
npm run test-client -- --grep "pages helpers"
```

### CODE REVIEW

File: client/my-sites/pages/helpers.js [hunk 1]
```diff
@@ -1,15 +1,63 @@
...
+		const insertChildren = ( pageID, indentLevel ) => {
+			const children = pagesByParent[ pageID ] || [];
+
+			forEach( children, child => {
+				const indentedChild = assign( child, {
+					indentLevel: indentLevel,
+				} );
+
+				sortedPages.push( indentedChild );
+				insertChildren( child.ID, indentLevel++ );
```
Comment (jblz) [inline client/my-sites/pages/helpers.js]: TODO: we should probably cap the recursion at some point.

File: client/my-sites/pages/helpers.js [hunk 2]
```diff
@@ -1,15 +1,63 @@
...
+			return result;
+		}, {} );
+
+		const sortedPages = [];
+
+		const insertChildren = ( pageID, indentLevel ) => {
+			const children = pagesByParent[ pageID ] || [];
+
+			forEach( children, child => {
+				const indentedChild = assign( child, {
```
Comment (jblz) [inline client/my-sites/pages/helpers.js]: TODO: don't mutate `child` object here.

File: client/my-sites/pages/helpers.js [hunk 3]
```diff
@@ -1,15 +1,63 @@
...
+					indentLevel: indentLevel,
```
Comment (jblz) [inline client/my-sites/pages/helpers.js]: TODO: use short literal syntax

File: client/my-sites/pages/helpers.js [hunk 4]
```diff
@@ -1,15 +1,59 @@
...
+		}, {} );
+
+		const sortedPages = [];
+
+		const insertChildren = ( pageID, indentLevel ) => {
+			const children = pagesByParent[ pageID ] || [];
+
+			forEach( children, child => {
+				sortedPages.push( assign( {}, child, { indentLevel } ) );
+				insertChildren( child.ID, indentLevel++ );
```
Comment (jblz) [inline client/my-sites/pages/helpers.js]: TODO: we should probably cap the recursion at some point.

**Comment by jblz** [comment]
I discovered a similar function in the [PostSelector](https://github.com/Automattic/wp-calypso/blob/85ce448bcb16f64d5c8b833611aaed36f18bdcd7/client/my-sites/post-selector/index.jsx#L58-L68) component used in the editor sidebar to show a hierarchical list of pages for choosing a page's parent.

That one is a bit more abstracted for any type of hierarchical post type & for various orderings than what is currently in this branch. 

If possible, we should converge on a single usage for both places & it should probably live in a new selector.

**Comment by mjuhasz** [comment]
Good job 👍

**Pull request merged**